src/cdomains/intDomain.ml

```diff
@@ -1853,27 +2395,156 @@ module IntDomTupleImpl = struct
-  let opt_map2 f = curry @@ function | Some x, Some y -> Some (f x y) | _ -> None
-  let map2  r (xa,xb,xc) (ya,yb,yc) = opt_map2 (r.f2  (module I1)) xa ya, opt_map2 (r.f2  (module I2)) xb yb, opt_map2 (r.f2  (module I3)) xc yc
-  let map2p r (xa,xb,xc) (ya,yb,yc) = opt_map2 (r.f2p (module I1)) xa ya, opt_map2 (r.f2p (module I2)) xb yb,  opt_map2 (r.f2p  (module I3)) xc yc
-  let to_list x = Tuple3.enum x |> List.of_enum |> List.filter_map identity (* contains only the values of activated domains *)
+  let to_list x = Tuple4.enum x |> List.of_enum |> List.filter_map identity (* contains only the values of activated domains *)
   let to_list_some x = List.filter_map identity @@ to_list x (* contains only the Some-values of activated domains *)
   let exists, for_all = let f g = g identity % to_list in List.(f exists, f for_all)
 
+
+
+  let refine_with_congruence ((a, b, c, d) : t) (cong : (int_t * int_t) option) : t=
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_congruence a cong
+    , opt I2.refine_with_congruence b cong
+    , opt I3.refine_with_congruence c cong
+    , opt I4.refine_with_congruence d cong )
+
+  let refine_with_interval (a, b, c, d) intv =
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_interval a intv
+    , opt I2.refine_with_interval b intv
+    , opt I3.refine_with_interval c intv
+    , opt I4.refine_with_interval d intv )
+
+  let refine_with_excl_list (a, b, c, d) excl =
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_excl_list a excl
+    , opt I2.refine_with_excl_list b excl
+    , opt I3.refine_with_excl_list c excl
+    , opt I4.refine_with_excl_list d excl )
+
+  let refine_with_incl_list (a, b, c, d) incl =
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_incl_list a incl
+    , opt I2.refine_with_incl_list b incl
+    , opt I3.refine_with_incl_list c incl
+    , opt I4.refine_with_incl_list d incl )
+
+
+    let mapp r (a, b, c, d) =
+    let map = BatOption.map in
+    ( map (r.fp (module I1)) a
+    , map (r.fp (module I2)) b
+    , map (r.fp (module I3)) c
+    , map (r.fp (module I4)) d)
+
+
+  let mapp2 r (a, b, c, d) =
+      BatOption.
+        ( map (r.fp2 (module I1)) a
+        , map (r.fp2 (module I2)) b
+        , map (r.fp2 (module I3)) c
+        , map (r.fp2 (module I4)) d )
+
+  let flat f x = match to_list_some x with [] -> None | xs -> Some (f xs)
+
+  let to_excl_list x = mapp2 { fp2 = fun (type a) (module I:S with type t = a and type int_t = int_t) -> I.to_excl_list } x |> flat List.concat
+  let to_incl_list x = mapp2 { fp2 = fun (type a) (module I:S with type t = a and type int_t = int_t) -> I.to_incl_list } x |> flat List.concat
```

Comment:
Why are the inclusion lists concatenated? If all the inclusion lists are a sound overapproximation, one could take the meet of the inclusion lists, no?
One would have to be careful to consider that some implementations of `to_incl_list` return `None` even when the abstract value is not bottom, but when it is only impractical to create the  complete inclusion list (e.g. the interval domain does this.)